From: Egmont Koblinger <egmont@gmail.com>
Date: Thu, 13 Jul 2023 21:59:29 +0200
Subject: widget: Invalidate ringview when the invalidating

When the ringview is not invalidated when the ring has changed leads to
failed assertion aborts when handling events, e.g. vte#2636, vte#2637,
vte#2632, vte#2577.

Bug: https://gitlab.gnome.org/GNOME/vte/-/issues/2636
Bug: https://gitlab.gnome.org/GNOME/vte/-/issues/2637
Bug-Debian: https://bugs.debian.org/1040049
Applied-upstream: 0.73.0, commit:461bc3e43c819fa0e3b62d0cf40ef533a69cc7f7
---
 src/vte.cc | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/vte.cc b/src/vte.cc
index b8e15d7..561cc42 100644
--- a/src/vte.cc
+++ b/src/vte.cc
@@ -2050,6 +2050,7 @@ Terminal::queue_adjustment_value_changed(double v)
         _vte_debug_print(VTE_DEBUG_ADJ,
                          "Scrolling by %f\n", dy);
 
+        m_ringview.invalidate();
         invalidate_all();
         match_contents_clear();
         emit_text_scrolled(dy);
@@ -2899,6 +2900,9 @@ Terminal::drop_scrollback()
         if (m_screen == &m_normal_screen) {
                 queue_adjustment_value_changed(m_normal_screen.insert_delta);
                 adjust_adjustments_full();
+                m_ringview.invalidate();
+                invalidate_all();
+                match_contents_clear();
         }
 }
 
@@ -7548,6 +7552,9 @@ Terminal::set_size(long columns,
 			gtk_widget_queue_resize(m_widget); // FIXMEgtk4?
 #endif
 
+                m_ringview.invalidate();
+                invalidate_all();
+                match_contents_clear();
 		/* Our visible text changed. */
 		emit_text_modified();
 	}
@@ -9780,6 +9787,10 @@ Terminal::set_scrollback_lines(long lines)
 	queue_adjustment_value_changed(scroll_delta);
 	adjust_adjustments_full();
 
+        m_ringview.invalidate();
+        invalidate_all();
+        match_contents_clear();
+
         return true;
 }
 
@@ -9962,7 +9973,9 @@ Terminal::reset(bool clear_tabstops,
         /* BiDi */
         m_bidi_rtl = FALSE;
 	/* Cause everything to be redrawn (or cleared). */
+	m_ringview.invalidate();
 	invalidate_all();
+	match_contents_clear();
 
         /* Reset XTerm window controls */
         m_xterm_wm_iconified = false;
